# PenTestAssignment
SLIIT Penetration Testing Assignment 2020 (3rd Semester)

by S.A.K.G. Samaraweera  (MS19804552)

## SYNOPSIS
This Git branch describes the SQL Injection attacks, could be exploited on Django framework. In order to setup environment please refer ReadMe.md of master branch.


## Pre-Requesits
Before we begin, we need to backup our database. So we can restore it later on, once completed this tutorial as you can use same database for the next tutorial.

### Backup Database
1. Switch to postgres user.
```
$ sudo -i -u postgres
```

2. Create backup file
```
$ pg_dump -u postgres hackmeifucan > before.bak
```

We can use below steps to restore our database once we completed this scenario.

### Restore database

```
$ sudo -i -u postgres
$ psql
> DROP DATABASE "hackmeifucan";
> CREATE DATABASE "hackmeifucan";
> GRANT ALL PRIVILEGES ON DATABASE hackmeifucan TO hackmeadmin;
> \q
$ psql
$ psql hackmeifucan < before.bak
```

## Exploit : CVE-2019-14234

Following Django vulnerability allows SQLInjection attacks using `django.contrib.postgres.fields.HStoreField` filed. `HStoreField` allows to store Dictory objects in Postgresql database.

First, we'll create "hstore" extention in the Postgresql database.

Access the database and execute below commands
```
$ sudo -i -u postgres
$ psql
> \c "hackmeifucan";
> CREATE EXRENSION hstore;
```

Once we've added extension, next we can try to migrate our vulnerable column.
Update your models.py file as below.

``` python
from django.db import models
from django.contrib.postgres.fields import HStoreField

class Student(models.Model):
    first_name = models.CharField(max_length=50)
    last_name = models.CharField(max_length=50)
    grade = models.CharField(max_length=100)
    # 'hstore' column
    marks = HStoreField(null=True, blank=True)
```
Now we can create our next migrations file.
```
$ python manage.py makemigrations

Migrations for 'StudentGradesApp':
  StudentGradesApp/migrations/0002_student_marks.py
    - Add field marks to student
```
In order to migrate, we have to add `HStoreExtension()` to the migration file.
``` python
from django.contrib.postgres.fields import HStoreField
# import HStoreExtension
from django.contrib.postgres.operations import HStoreExtension
from django.db import migrations

class Migration(migrations.Migration):

    dependencies = [
        ('StudentGradesApp', '0001_initial'),
    ]

    operations = [
        # add HStoteExtension()
        HStoreExtension(),

        migrations.AddField(
            model_name='student',
            name='marks',
            field=HStoreField(blank=True, null=True),
        ),
    ]
```
We can now do the database migration. Execute below command to migrate.
```
$ python manage.py migrate
```

## Scan vulnerability

First, we'll create new Student entity in database using Django shell prompt. We can do the same thing using Django admin panel or implementing a Django view. But for sake of clarity we'll proceed this method. Eitherway we have to create new instance of Student model and persist in database.

```
$ python manage.py shell
>>>
```
Now. we'll create an the Student entity.

```
...
>>> from StudentGradesApp.models import Student
>>> Student.objects.create(first_name="John", last_name="Doe",grade="95",marks= {'maths':"88",'science':"75"})
<Student: Student object (2)>
>>>

```
Then, we'll try to query the created Student entity from database. Django calls `Queryset.filter()` method for this.
```
>>> Student.objects.filter(marks_contains='maths')
```
In above query, I intentionaly made a mistake. Closely look in our query stack trace. You'll notice some SQL syntax are directly embeded in to our model fields.

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/models/query.py", line 250, in __repr__
    data = list(self[:REPR_OUTPUT_SIZE + 1])
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/models/query.py", line 274, in __iter__
    self._fetch_all()
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/models/query.py", line 1242, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/models/query.py", line 55, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1140, in execute_sql
    cursor.execute(sql, params)
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 99, in execute
    return super().execute(sql, params)
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/admin/django/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.InternalError: Unexpected end of string
LINE 1: ...dent" WHERE "StudentGradesApp_student"."marks" @> 'maths'  L...
```
Now we'll try to exploit this.